src/Autofac/Core/Activators/ProvidedInstance/ProvidedInstanceActivator.cs

```diff
@@ -64,23 +65,60 @@ private object GetInstance()
         /// </summary>
         public bool DisposeInstance { get; set; }
 
-        /// <summary>
-        /// Releases unmanaged and - optionally - managed resources.
-        /// </summary>
-        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        /// <inheritdoc />
         protected override void Dispose(bool disposing)
         {
             // Only dispose of the instance here if it wasn't activated. If it was activated,
             // then either the owning lifetime scope will dispose of it automatically
             // (see InstanceLookup.Activate) or an OnRelease handler will take care of it.
-            if (disposing && DisposeInstance && _instance is IDisposable disposable && !_activated)
+            if (disposing && DisposeInstance && !_activated)
             {
-                disposable.Dispose();
+                // If we are in synchronous dispose, and an object implements IDisposable,
+                // then use it.
+                if (_instance is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
+                else if (_instance is IAsyncDisposable)
+                {
+                    // Type only implements IAsyncDisposable, which is not valid if there
+                    // is a synchronous dispose being done.
+                    throw new InvalidOperationException(string.Format(
+                        DisposerResources.Culture,
+                        DisposerResources.TypeOnlyImplementsIAsyncDisposable,
+                        _instance.GetType().FullName));
+                }
             }
 
             base.Dispose(disposing);
         }
 
+        /// <inheritdoc />
+        protected override async ValueTask DisposeAsync(bool disposing)
+        {
+            if (disposing && DisposeInstance && !_activated)
+            {
+                // If the item implements IAsyncDisposable we will call its DisposeAsync Method.
+                if (_instance is IAsyncDisposable asyncDisposable)
+                {
+                    var vt = asyncDisposable.DisposeAsync();
```

Comment:
This little block of "if/else" plus "get value task, check completed, await result if not" appears a few times. Would it be worth having a static method for this with `[MethodImpl(MethodImplOptions.AggressiveInlining)]` on it to force it to be inlined? I think it should compile down to the same thing but maybe reduce some duplication? Or maybe it's not worth the lack of reading clarity. Hmmm. I don't feel strongly about it, figured I'd raise it as a possibility.

Reply:
I get kind of nervous whenever I write code that passes a `ValueTask` as a parameter. The guidance from Microsoft is pretty clear that ValueTasks should generally just be [immediately awaited, or have the IsCompleted check](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/#valid-consumption-patterns-for-valuetasks), and that's it.  Not saying it's impossible trying to write a method that makes this check common, but it seems to go against best practice.

It seems more verbose, but I think the controlled management of the ValueTask, plus the readability, outweighs the duplication.

Reply:
I'll buy that.